### PR TITLE
Support nested objects in schemas

### PR DESCRIPTION
**Pull request by shireesh-illumio**

### CODE REVIEW

**Comment by gshireesh** [comment]
double check logic to see if nested object has some fields immutable and other modifiable.
how the update mask and req are working.

**Comment by shireesh-illumio** [comment]
the diff is not at the attribute level. Even if a single element changed in the object its passed down as change in update request. and that needs to be handled by the config server. this helps because we get the full expected state so we can apply the changes.

File: api/providergen/providergen_test.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+	data := providerTemplateData{
+		Package:               "testpkg",
+		ProviderTypeName:      "Provider",
+		Models:                make([]model, 0),
+		NewRequestFuncs:       make([]convertFunc, 0, countCount*3),
+		NewUpdateRequestFuncs: make([]convertFunc, 0, countCount),
+		CopyResponseFuncs:     make([]convertFunc, 0, countCount*3),
+		Resources:             make([]resourceData, 0, countCount),
+	}
```
Comment (rlenglet) [inline api/providergen/providergen_test.go:85]: Test the full generated code.

File: api/providergen/providergen_test.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	countCount := 1
+	data := providerTemplateData{
+		Package:               "testpkg",
+		ProviderTypeName:      "Provider",
+		Models:                make([]model, 0),
+		NewRequestFuncs:       make([]convertFunc, 0, countCount*3),
+		NewUpdateRequestFuncs: make([]convertFunc, 0, countCount),
+		CopyResponseFuncs:     make([]convertFunc, 0, countCount*3),
+		Resources:             make([]resourceData, 0, countCount),
+	}
```
Comment (rlenglet) [inline api/providergen/providergen_test.go:154]: Test the full generated code.

**Comment by rlenglet** [comment]
@shireesh-illumio any update on this?

File: api/protogen/protogen.go [hunk 3]
```diff
@@ -267,15 +260,60 @@ func terraformAttributeTypeToProtoType(attrName string, attrType attr.Type) (rep
...
 
+// Converts a Terraform object attribute to a Protocol Buffer message type.
+func terraformObjectAttributeTypeToProtoType(attrName string, attrTypes map[string]attr.Type) (repeated bool, protoType string, messages *message, err error) {
+	messageName := schema.ProtoMessageName(attrName)
+	newMessage := &message{
+		Name:   messageName,
+		Fields: []field{},
+	}
+
+	for name, attrType := range attrTypes {
```
Comment (rlenglet) [inline api/protogen/protogen.go:292]: The order of items when iterating over a map is deterministic.
You must make sure that the behavior of this function is always deterministic.

File: api/protogen/protogen.go [hunk 4]
```diff
@@ -267,15 +260,60 @@ func terraformAttributeTypeToProtoType(attrName string, attrType attr.Type) (rep
...
+		repeated, t, msg, err := terraformAttributeTypeToProtoType(name, attrType)
+		if err != nil {
+			return false, "", nil, fmt.Errorf("failed to parse field %s in object %s: %w", name, attrName, err)
+		}
+
+		newMessage.Fields = append(newMessage.Fields, field{
+			Repeated: repeated,
+			Type:     t,
+			Name:     name,
+			Tag:      len(newMessage.Fields) + 1,
```
Comment (rlenglet) [inline api/protogen/protogen.go:302]: You must use the `apiSpecTagger` to generate tags to preserve previously allocated tags, to preserve backward compatibility on the wire when adding or remove fields.

**Comment by gshireesh** [comment]
fixed

**Comment by rlenglet** [comment]
For the record, this PR was closed and cloned into #124.

**Pull request closed without merging**